onnxruntime/core/optimizer/dynamic_quantize_matmul_fusion.cc

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "dynamic_quantize_matmul_fusion.h"
+
+#include "core/optimizer/initializer.h"
+#include "core/graph/graph_utils.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+/**
+DynamicQuantizeMatMulFusion will fuse subgraph like below into DynamicQuantizeMatMul:
+    (input)
+       |
+       v
+DynamicQuantizeLinear --------+
+       |                      |
+       v                      v
+MatMulInteger (B const)      Mul (B const)
+       |                      |
+       v                      v
+     Cast ------------------>Mul
+                              |
+                              v
+                           (output)
+*/
+Status DynamicQuantizeMatMulFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  std::vector<std::reference_wrapper<Node>> nodes_to_remove;
+
+  for (auto node_index : node_topology_list) {
+    auto* node_ptr = graph.GetNode(node_index);
+    if (nullptr == node_ptr)
+      continue;  // node was removed
+
+    auto& mul_node = *node_ptr;
+
+    ORT_RETURN_IF_ERROR(Recurse(mul_node, modified, graph_level, logger));
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(mul_node, "Mul", {7}) ||
+        !graph_utils::IsSupportedProvider(mul_node, GetCompatibleExecutionProviders())) {
+      continue;
+    }
+
+    // Left Parents path
+    std::vector<graph_utils::EdgeEndToMatch> left_parent_path{
+        {0, 0, "Cast", {6, 9}, kOnnxDomain},
+        {0, 0, "MatMulInteger", {10}, kOnnxDomain},
+        {0, 0, "DynamicQuantizeLinear", {11}, kOnnxDomain}};
+
+    std::vector<graph_utils::EdgeEndToMatch> right_parent_path{
+        {0, 1, "Mul", {7}, kOnnxDomain},
+        {1, 0, "DynamicQuantizeLinear", {11}, kOnnxDomain}};
+
+    std::vector<const Node::EdgeEnd*> left_edges;
+    std::vector<const Node::EdgeEnd*> right_edges;
+    if (!graph_utils::FindPath(mul_node, true, left_parent_path, left_edges, logger) ||
+        !graph_utils::FindPath(mul_node, true, right_parent_path, right_edges, logger)) {
+      continue;
+    }
+
+    Node& cast_node = const_cast<Node&>(left_edges[0]->GetNode());
+    Node& matmulinteger_node = const_cast<Node&>(left_edges[1]->GetNode());
```

Comment:
You can avoid the ugly const_cast by using graph.GetNode(left_edges[0]->GetNode().Index())

Or if it's a common sequence to call FindPath and modify the returned nodes, perhaps FindPath could be updated or have a variant that returns mutable nodes